drivers/wifi/esp32/src/esp_wifi_drv.c

```diff
@@ -153,37 +231,181 @@ static void esp_wifi_event_task(void)
+
+	if (ret != ESP_OK) {
+		LOG_INF("Failed to disconnect from hotspot");
+	}
+
+	return ret;
+}
+
+static int esp32_wifi_connect(const struct device *dev,
+			    struct wifi_connect_req_params *params)
+{
+	struct esp32_wifi_runtime *data = dev->data;
+	int ret;
+
+	if (data->state == ESP32_STA_CONNECTING || data->state == ESP32_STA_CONNECTED) {
+		wifi_mgmt_raise_connect_result_event(esp32_wifi_iface, -1);
+		return -EALREADY;
+	}
+
+	if (data->state != ESP32_STA_STARTED) {
+		LOG_ERR("WiFi not in station mode");
+		wifi_mgmt_raise_connect_result_event(esp32_wifi_iface, -1);
+		return -EIO;
+	}
+
+	data->state = ESP32_STA_CONNECTING;
+
+	wifi_config_t wifi_config;
+
+	memset(&wifi_config, 0, sizeof(wifi_config_t));
+
+	memcpy(wifi_config.sta.ssid, params->ssid, params->ssid_length);
+	wifi_config.sta.ssid[params->ssid_length] = '\0';
+
+	if (params->security == WIFI_SECURITY_TYPE_PSK) {
+		memcpy(wifi_config.sta.password, params->psk, params->psk_length);
+		wifi_config.sta.password[params->psk_length] = '\0';
+		wifi_config.sta.threshold.authmode = WIFI_AUTH_WPA2_PSK;
+	} else {
+		wifi_config.sta.threshold.authmode = WIFI_AUTH_OPEN;
+	}
+
+	wifi_config.sta.pmf_cfg.capable = true;
+	wifi_config.sta.pmf_cfg.required = false;
+
+	ret = esp_wifi_set_config(ESP_IF_WIFI_STA, &wifi_config);
+	ret |= esp_wifi_set_mode(WIFI_MODE_STA);
+	ret |= esp_wifi_connect();
+
+	return ret;
+}
+
+static int esp32_wifi_scan(const struct device *dev, scan_result_cb_t cb)
+{
+	struct esp32_wifi_runtime *data = dev->data;
+	int ret = 0;
+
+	if (data->scan_cb != NULL) {
+		LOG_INF("Scan callback in progress");
+		return -EINPROGRESS;
+	}
+
+	data->scan_cb = cb;
+
+	wifi_scan_config_t scan_config = { 0 };
+
+	ret = esp_wifi_set_mode(WIFI_MODE_STA);
+	ret |= esp_wifi_scan_start(&scan_config, false);
+
+	if (ret != ESP_OK) {
+		return -EIO;
+	}
+
+	return 0;
+};
+
+static int esp32_wifi_ap_enable(const struct device *dev,
+			 struct wifi_connect_req_params *params)
+{
+	struct esp32_wifi_data *data = dev->data;
+
+	/* Build WiFi configuration for AP mode */
+	wifi_config_t wifi_config = {
+		.ap = {
+			.max_connection = 5,
+		},
+	};
+
+	strncpy((char *) wifi_config.ap.ssid, params->ssid, params->ssid_length);
+
+	if (params->psk_length == 0) {
+		memset(wifi_config.ap.password, 0, sizeof(wifi_config.ap.password));
+		wifi_config.ap.authmode = WIFI_AUTH_OPEN;
+	} else {
+		strncpy((char *) wifi_config.ap.password, params->psk, params->psk_length);
+		wifi_config.ap.authmode = WIFI_AUTH_WPA_WPA2_PSK;
+	}
+
+	/* Start WiFi in AP mode with configuration built above */
+	esp_err_t err = esp_wifi_set_mode(WIFI_MODE_AP);
+
+	if (err != ESP_OK) {
+		LOG_ERR("Failed to set WiFi mode: %d", err);
+		return err;
+	}
+	err = esp_wifi_set_config(WIFI_IF_AP, &wifi_config);
+	if (err != ESP_OK) {
+		LOG_ERR("Failed to set WiFi config: %d", err);
+		return err;
+	}
+	err = esp_wifi_start();
+	if (err != ESP_OK) {
+		LOG_ERR("Failed to start WiFi: %d", err);
+		return err;
```

Comment:
Error codes should be mapped to posix error codes, as application layer will depend on those.